getpaid/views.py

```diff
@@ -33,57 +29,77 @@ def get(self, request, *args, **kwargs):
-            context = self.get_context_data()
-            context['gateway_url'] = processor.get_gateway_url(self.request)[0]
-            context['form'] = processor.get_form(gateway_url_tuple[2])
+        if method.upper() == 'GET':
+            if params:
+                url = "{url}?{params}".format(url=url, params=urlencode(params))
+            return http.HttpResponseRedirect(url)
+        elif method.upper() == 'POST':
+            context = self.get_context_data(
+                form=payment.get_form(params),
+                gateway_url=url
+            )
 
             return TemplateResponse(
                 request=self.request,
-                template=self.get_template_names(),
+                # template=self.get_template_names(),
+                template=payment.get_template_names(view=self),
                 context=context)
         else:
-            raise ImproperlyConfigured()
+            raise exceptions.ImproperlyConfigured('Only GET and POST supported.')
 
     def form_invalid(self, form):
-        raise PermissionDenied
+        raise exceptions.PermissionDenied
 
 
 class FallbackView(RedirectView):
+    """
+    FallbackView (in form of either SuccessView or FailureView) handles the
+    return from payment broker.
+    """
     success = None
     permanent = False
 
-    def get_redirect_url(self, **kwargs):
-        from getpaid.models import Payment
-        self.payment = get_object_or_404(Payment, pk=self.kwargs['pk'])
+    def get_redirect_url(self, *args, **kwargs):
+        Payment = swapper.load_model('getpaid', 'Payment')
 
+        getpaid_settings = getattr(settings, 'GETPAID', {})
+        payment = get_object_or_404(Payment, pk=self.kwargs['pk'])
         if self.success:
-            url_name = getattr(settings, 'GETPAID_SUCCESS_URL_NAME', None)
+            payment.on_success()
+            url = getattr(getpaid_settings, 'SUCCESS_URL', None)
         else:
-            url_name = getattr(settings, 'GETPAID_FAILURE_URL_NAME', None)
+            payment.on_failure()
+            url = getattr(getpaid_settings, 'FAILURE_URL', None)
+
+        if url is not None:
+            return resolve_url(url, pk=payment.order.pk)  # we may want to return to the Order summary or smth
+        return resolve_url(payment.order.get_redirect_url(payment, success=self.success))
+
+
+class SuccessView(FallbackView):
+    success = True
+
+
+class FailureView(FallbackView):
+    success = False
+
+
+class CallbackView(View):
+    def post(self, request, pk, *args, **kwargs):
+        Payment = swapper.load_model('getpaid', 'Payment')
+        payment = get_object_or_404(Payment, pk=pk)
+        return payment.handle_callback(request, *args, **kwargs)
 
-        if url_name is not None:
-            return reverse(url_name, kwargs={'pk': self.payment.order_id})
-        return self.payment.order.get_absolute_url()
+    def get(self, request, pk, *args, **kwargs):
```

Comment:
Shouldn't this be redirect? GET request, has a lot of potential to be cached on different layer, while POST request indicates some backend action. Redirect make sure, that someone doesn't necessarily use this endpoint wrong.

I think (but there is a chance that I maybe am wrong on that), but most providers are doing webhook notifications via POST request, like Stripe for example.

Also for dummy backend, I think this will fail, because handle_callback will try to load JSON which isn't there.

Reply:
Hm... Major version bump allows breaking old APIs so I really should remove this :smile: Thanks :+1: